File: python/299_BullsAndCows.py
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        bulls, cows = 0, 0
        bull_indices, secret_dict = set(), {}
        for l in secret: secret_dict[l] = secret_dict[l] + 1 if l in secret_dict else 1
        
        # bulls
        for i in range(len(guess)):
            if guess[i] == secret[i]: 
                bulls += 1
                secret_dict[guess[i]] -= 1
                bull_indices.add(i)
                
        #cows
        for i in range(len(guess)):
            if guess[i] in secret_dict and secret_dict[guess[i]] > 0 and i not in bull_indices: 
                cows += 1
                secret_dict[guess[i]] -= 1
        
        return f'{bulls}A{cows}B'

File: swift/299_BullsAndCows.swift
class Solution {
    func getHint(_ secret: String, _ guess: String) -> String {
        var counts: [Character : Int] = [Character : Int]()
        
        var bulls: Int = 0, cows: Int = 0
        
        var index = secret.startIndex
        while index != secret.endIndex {
            var s: Character = secret[index], g: Character = guess[index]
            if s == g {
                bulls += 1
            } else {
                if counts[s, default: 0] < 0 { cows += 1 }
                if counts[g, default: 0] > 0 { cows += 1 }
                
                counts[s, default: 0] += 1
                counts[g, default: 0] -= 1
            }
            secret.formIndex(after: &index)
        }
        return "\(bulls)A\(cows)B"
    }
}
